### DISCUSSION TITLE
Help with ui.download and mounted storage on Posit Connect

### DISCUSSION THREAD

**Discussion by zilch42** [zauberzeug/nicegui#2386] [Q&A] [answered]
### Question

I'm having some trouble getting a ui.download to work with a file in mounted storage on posit connect. 

```python
from fastapi import FastAPI
from nicegui import ui
import os


with ui.column():
    path = "/mounted_storage/work/sip-bibliomet/work/project/dashboard_datasets/dataset.xlsx"
    ui.label("Path Exists:")
    ui.label(str(os.path.exists(path)))
    ui.button('Download dataset', on_click=lambda: ui.download(path))


FastApp = FastAPI()
ui.run_with(FastApp)

if __name__ == '__main__':
    print('Please start with "uvicorn main:FastApp --reload"')
```

The file I'm trying to download is in a folder on a mounted storage drive on the deployment server. The app can definitely access that location because  `os.path.exists(path)` returns `True` and my larger app is loading a dataset from that folder successfully. I can see the file it should be downloading present in that folder. 

When I click download I get a new empty tab and the following message in the log:
```
 http://127.0.0.1:40025/app_name/_nicegui/auto/static/d3e390abbbd942abdb0aeb0801dcb67d/dataset.xlsx not found
```

If I just put a file in the root app directory next to `main.py` and try and download it from there, that seems to work fine so it seems to be an issue with `ui.download` and the mounted storage but I'm not sure why. Am I doing something wrong? 

This is using nicegui 1.4.9 as I can't currently test on 1.4.11 on posit connect (see #2385)

**Comment by rodja** [comment]
Just a wild guess: maybe you need to limit to one worker? See https://github.com/zauberzeug/nicegui/issues/469#issuecomment-1474698426 for a issue which could be solved like this.

**Reply by zilch42** [reply]
I'm deploying with the following code which doesn't have a --workers argument
```
rsconnect deploy fastapi -n rconnect --entrypoint dashboard_main:FastApp .
```

But I think the 'Max processes' parameter in the runtime settings on Posit Connect is probably the same thing?
![image](https://github.com/zauberzeug/nicegui/assets/39004650/75674e7d-4134-40cd-a622-2a436b7fcb70)
I've set that to 1 and that hasn't solved it.

**Comment by rodja** [comment]
Hmm... just wondered why the url path starts with `http://127.0.0.1:40025/app_name`. Should that not be some public available domain and a real identifier? Or is your app called "app_name"? URL prefix support for `ui.download` was added as part of https://github.com/zauberzeug/nicegui/pull/2250 (released in 1.4.9).

**Reply by zilch42** [reply]
`app_name` has been changed for privacy but the `http://127.0.0.1:40025` is real. I should probably be clearer about which bits I've censored. The bit in the curly braces below has been changed. 
```
http://127.0.0.1:40025/{app_name}/_nicegui/auto/static/d3e390abbbd942abdb0aeb0801dcb67d/dataset.xlsx not found
```

But yes I did wonder if it was odd that it seems to be pointing to a local address on the server. 

I'm viewing the app on a url that looks like this
```
https://rconnect.it.{companydomain.com}/connect/#/apps/b911095a-5797-429e-ace9-XXXXXXXXXX/
```

> URL prefix support for `ui.download` was added as part of #2250 (released in 1.4.9).

That made me curious so I went back to 1.4.8 (I thought I did have the dataset download working at some point) but that didn't work either.

**Reply by zilch42** [reply]
A little more information: on 1.4.8, rather than getting an new empty tab and a message in the log, I get a new tab with the following url and a message saying "Firefox Can't Open This Page"

```
https://rconnect.it.{companydomain.com}/_nicegui/auto/static/fa971c61e18b7c7f25567c0dc75211bc/dataset.xlsx
```

So on 1.4.8 I'm getting the same domain that I'm viewing the app on, while on 1.4.9 I'm getting 127.0.0.1...
Neither actually download the fine though

**Reply by rodja** [reply]
Are images displayed correctly [when using a local path](https://nicegui.io/documentation/image#local_files)? What's the url of such an image? It looks like Posit Connect is not doing unexpected url rewriting. If your url is `https://rconnect.it.{companydomain.com}/connect/#/apps/b911095a-5797-429e-ace9-XXXXXXXXXX/` NiceGUI assumes the proxy is setting the `X-Forwarded-Prefix`-header to `/connect/#/apps/b911095a-5797-429e-ace9-XXXXXXXXXX/`:
https://github.com/zauberzeug/nicegui/blob/ccab1812fd414080f0f844827b2f7d8ea1daf77f/nicegui/client.py#L116

From the docs at https://docs.posit.co/ide/server-pro/access_and_security/running_with_a_proxy.html I assume they are using Nginx. Maybe have a look at https://github.com/zauberzeug/nicegui/tree/main/examples/nginx_subpath/ for an example how Nginx needs to be configured to properly support the subpath.

**Reply by zilch42** [reply]
The plot thickens...

> Are images displayed correctly [when using a local path](https://nicegui.io/documentation/image#local_files)? What's the url of such an image?

Yes. 

```python
from fastapi import FastAPI
from nicegui import ui
import os


with ui.column():
    path_dataset = "/{mounted_storage}/dashboard_datasets/dataset.xlsx"
    path_logo = "/{mounted_storage}/dashboard_datasets/logo.png"
    ui.label("Dataset Exists:")
    ui.label(str(os.path.exists(path_dataset)))
    ui.label("Image Exists:")
    ui.label(str(os.path.exists(path_logo)))

    ui.button('Download dataset', on_click=lambda: ui.download(path_dataset))
    ui.button('Download image', on_click=lambda: ui.download(path_logo))

    with ui.row():
        ui.label("Image in app folder:")
        ui.image("logo.png")
    with ui.row():
        ui.label("Image on local path:")
        ui.image(path_logo)


FastApp = FastAPI()
ui.run_with(FastApp)

if __name__ == '__main__':
    print('Please start with "uvicorn dashboard_main:FastApp --reload"')
```

In that example, both the logo in the app folder next to `main.py` and the one sourced from the mounted path are loaded properly and get a similar url.

```
# ui.image("logo.png")
https://rconnect.it.{company.com}/{app_name}/_nicegui/auto/static/6618bb241f670c2dc83ae272649e649d/logo.png
# ui.image(path_logo)
https://rconnect.it.{company.com}/{app_name}/_nicegui/auto/static/b4d0668a8c7f7dbc3588ea25590bc238/logo.png
```

What is really interesting is that downloading the logo from the mounted path with `ui.download` WORKS, while downloading the xlsx file does not, BUT, download of the .png file only works if it has been displayed on the page with `ui.image` first. If I comment out those lines, then `ui.download` of the logo from the mounted storage just opens an empty tab. 

So it would seem that displaying the image from the mounted storage makes it available for `ui.download` to access. Maybe there is something I need to do to make the .xlsx file available too?

> Maybe have a look at https://github.com/zauberzeug/nicegui/tree/main/examples/nginx_subpath/ for an example how Nginx needs to be configured to properly support the subpath.

Is this something I would do on my app's side or I would have to contact my system admins about that? It initially looks beyond my current level of understanding.

**Reply by rodja** [reply]
Ok. To make files available there is `app.add_static_file` which is internally used by `ui.image` but also by `ui.download`. I wonder what happens if you do a manual `app.add_static_file(path_to_dataaset)` or `app.add_static_file(path_to_dataaset, single_use=True)`...

For Nginx config mit Posit you would need to consult their docs. But if image is working, I assume Path-Prefix is already correct.

**Reply by zilch42** [reply]
Thanks rodja. So if I just add `app.add_static_file(path_dataset)` somewhere at the top of my code, that works. If I use `single_use` then it works the first time but it's a bit hit and miss (navigate away, then back and its gone).

```python
from fastapi import FastAPI
from nicegui import ui, app
import os

with ui.column():
    path_dataset = "/{mounted_storage}/dashboard_datasets/dataset.xlsx"
    path_logo = "/{mounted_storage}/dashboard_datasets/logo.png"
    app.add_static_file(local_file=path_dataset, single_use=True)
    ui.label("Dataset Exists:")
    ui.label(str(os.path.exists(path_dataset)))
    ui.label("Image Exists:")
    ui.label(str(os.path.exists(path_logo)))

    ui.button('Download dataset', on_click=lambda: trigger_download(path_dataset))
    ui.button('Download image', on_click=lambda: trigger_download(path_logo))

    with ui.row():
        ui.label("Image in app folder:")
        ui.image("logo.png")

FastApp = FastAPI()
ui.run_with(FastApp)

if __name__ == '__main__':
    print('Please start with "uvicorn dashboard_main:FastApp --reload"')
```
However, I don't really want to add every possible dataset as a static file, 1 because there may be a lot of them, and 2 because I don't necessarily want every dataset accessible to everyone. 

I thought maybe I could just add the static file before download with a little helper function like this
```python
def trigger_download(file_path: str):
    app.add_static_files(local_file=file_path)
    ui.download(file_path)

ui.button('Download dataset', on_click=lambda: trigger_download(path_dataset))
```

but for whatever reason that doesn't work, with or without `single_use`.

**Reply by falkoschindler** [reply, marked as answer]
@zilch42 If you want to use `ui.download` with `app.add_static_files`, you need to call `ui.download` with a route instead of a local filepath. Otherwise the static route is ineffective and `ui.download` creates a new single-use route.

**Reply by zilch42** [reply]
Ah thanks @falkoschindler and @rodja. That has solved it for my simple example. I'm still having an issue in my larger app but it miight have a different root and I don't have a reproducible example yet
